docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,505 @@
+    }
+    ```
+
+#### Validations
+##### ClusterClass
+- For object creation:
+  - `spec.worker.deployments[i].class` field must be unique within a ClusterClass.
+- For object updates:
+  - `spec.worker.deployments` supports adding new deployment classes.
+
+##### Cluster
+- For object creation:
+  - spec.managed and spec.controlPlaneRef cannot be simultaneously set.
+  - If spec.managed.class is set, spec.managed.version cannot be empty and must be a valid semver.
+
+- For object updates:
+  - If spec.managed.class is in use, it cannot be unset or modified.
+  - spec.managed.version cannot be unset and must be a valid semver, if being updated.
+  - `spec.managed.version` cannot be downgraded.
+  - `spec.managed.controlPlane.replicas` can be updated.
+  - `spec.managed.worker.deployments[i].replicas` can be updated.
+  - `spec.managed.worker.deployments[i].name` field must be unique within a Cluster
+  - A set of worker nodes can be added to or removed from the `spec.class.managed.worker.deployments` list.
+
+#### Behaviors
+This section lists out the behavior for Cluster objects using `ClusterClass` in case of creates and updates.
+
+##### Create a new Cluster using ClusterClass object
+1. User creates a ClusterClass object.
+   ```yaml
+    apiVersion: cluster.x-k8s.io/v1alpha4
+    kind: ClusterClass
+    metadata:
+      name: mixed
+      namespace: bar
+    spec:
+      controlPlane:
+        ref:
+          apiVersion: controlplane.cluster.x-k8s.io/v1alpha4
+          kind: KubeadmControlPlaneTemplate
+          name: vsphere-prod-cluster-template-kcp
+      worker:
+        deployments:
+        - class: linux-worker
+          template:
+            bootstrap:
+              ref:
+                apiVersion: bootstrap.cluster.x-k8s.io/v1alpha4
+                kind: KubeadmConfigTemplate
+                name: existing-boot-ref
+            infrastructure:
+              ref:
+                apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+                kind: VSphereMachineTemplate
+                name: linux-vsphere-template
+        - class: windows-worker
+          template:
+            bootstrap:
+              ref:
+                apiVersion: bootstrap.cluster.x-k8s.io/v1alpha4
+                kind: KubeadmConfigTemplate
+                name: existing-boot-ref-windows
+            infrastructure:
+              ref:
+                apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+                kind: VSphereMachineTemplate
+                name: windows-vsphere-template
+      infrastructure:
+        ref:
+          apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+          kind: VSphereClusterTemplate
+          name: vsphere-prod-cluster-template
+   ```
+2. User creates a cluster using the class name and defining the managed topology.
+   ```yaml
+    apiVersion: cluster.x-k8s.io/v1alpha4
+    kind: Cluster
+    metadata:
+      name: foo
+      namespace: bar
+    spec:
+      managed:
```

Comment:
What about just calling it `topology:`? The topology itself is managed, not the cluster. Calling a cluster "managed" just implies other things and I really think we should stay away from that. `topology:` would imply, "this is what my cluster's topology should look like". 

Anyone else has suggestions?

cc @puja108

Reply:
Naming is hard

What I like of the current name is that it reflects the fact that everything is under Cluster.Managed is actually controlled (aka managed) by the Cluster CR/Cluster controller in Cluster API.

Instead, considering that user is still free to add more MachineDeployments/Pools manually and self-manage them as of today, I find the use of the term topology a little bit trickier, because in my mind such mixed scenario requires some more nuances like: 
- The "cluster topology" including both MachineDeployments managed at Cluster level and the MachineDeployments/Pools manually added/self managed.
- The "managed topology" of a cluster, including only MachineDeployments managed at Cluster level
- The "self-managed topology" of a cluster, including only MachineDeployments/Pools manually added

However I agree, that the downside of "managed" is the potential confusion with AKS, EKS, GKE, which are Kubernetes offering managed by something/someone outside of Cluster API.

Reply:
HUGE +1 from me as well. We really can't start referring to clusters built/defined with this new `ClusterClass` configuration by a different name (e.g., "managed cluster" vs "cluster"). In this particular case, the term "managed cluster" is too general, and in the cluster-api ecosystem it is very sensible right now to refer to *any* cluster-api workload cluster as a "managed cluster" — these cluster *are managed* by the cluster-api _management cluster_.

And, more broadly, this seems to be a natural evolution of the cluster-api abstraction. We should expect there to be many more in the future. We don't want to back ourselves into a corner by having to create a distinct terminology for every evolving "type" of cluster-api cluster configuration.

My 2¢ is that we should try to always refer to this in language as "a cluster that uses ClusterClass" to reflect that this is just one possible configuration paradigm for a cluster-api cluster, one that reflects a certain vintage ("I built my cluster from cluster-api version XYZ and am taking advantage of the latest functionality!") and operational preference. Rather than "a discrete _type_ of cluster-api cluster".

Reply:
I agree with the concern that the term **Managed Cluster** is already in use outside of the context of CAPI, and overloading the use of the term will be misleading. 
But does naming the field `managed` imply that the cluster is managed? IMO it just states that the spec described within this field is managed and not necessarily that the cluster is managed. The confusion could be arising since the golang struct is named as `ManagedCluster` and I vote for updating that name to `ManagedTopology` to alleviate the confusion and keeping the name of the field as `managed`.